### DISCUSSION TITLE
2FA without mobile device

### DISCUSSION THREAD

**Discussion by Anpanator** [community/community#49715] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

I understand GitHub wants to require 2FA soon and in principle I think this is a good thing.

However, for various reasons, I cannot use SMS or Authenticator 2FA. I understand WebAuthn is available, but only as a fallback.

Is it possible to use WebAuthn as the main 2FA method? If not, why? Services like Twitter and others seem to allow this.

I have a bunch of security keys here that I could use (and I always enroll at least 2 for backup), but not a mobile device.

I found https://github.com/orgs/community/discussions/23490, but it's almost 5 years old, I'm wondering if anything changed here?

I appreciate any help, as once this is required, I'd have to migrate projects elsewhere otherwise.

**Comment by ghost** [comment]
Hi there! I understand that you're looking for alternative options for 2FA on GitHub, since you cannot use SMS or Authenticator 2FA. I'm happy to help you explore some possible solutions.

It's true that GitHub currently does not offer WebAuthn as the main 2FA method, but it is available as a fallback option. However, you can still use your security keys for 2FA on GitHub. In fact, security keys are a recommended 2FA method by GitHub.

To set up security key 2FA on GitHub, you can follow these steps:

1. Go to your GitHub account settings and click on "Security."
2. Under "Two-factor authentication," click "Set up two-factor authentication."
3. Follow the prompts to choose "Security key" as your 2FA method.
4. Insert your security key and follow the prompts to complete the setup.

Once you've set up security key 2FA, you can use your security keys to log in to GitHub without the need for a mobile device.

I hope this helps you find a solution that works for you! Let me know if you have any further questions or if there's anything else I can assist you with.

**Reply by Anpanator** [reply]
Thanks for your answer, unfortunately, it is not a solution, as to enable 2FA to begin with, you need either of the two options I mentioned initially. WebAuthn is not possible:
![image](https://user-images.githubusercontent.com/11788825/224361611-4e8ddde3-cd0c-49ca-814a-e81315f14130.png)

**Reply by ghost** [reply]
Hey, I've got a suggestion for you! If you're having trouble with 2FA options like SMS and WebAuthn, you might want to try using the [Authenticator Chrome Extension](https://chrome.google.com/webstore/detail/authenticator/bhghoamapcdpbohphigoooaddinpkbai). I use it myself and find it to be really reliable. Give it a try and see if it works well for you too!

**Reply by Anpanator** [reply]
I'll note that down as a possible workaround, however it's not a great solution, as it relies on some arbitrary software. The benefit of WebAuthn is that it is supported natively by all major browsers (and if that is compromised, you are doomed in any case).

Thank you for the hint though, I appreciate that.

**Reply by ghost** [reply]
Thank you for considering the suggestion. I understand your concerns about relying on third-party software for authentication, and it's great to hear that you value the native support of WebAuthn in major browsers. I hope you find a solution that works well for you and meets your security needs.

**Reply by gms2009** [reply]
I think github should put out a step by step instruction for firefox on linux to enable 2FA,  too difficult to find a way to do it.

**Reply by GarthWilson** [reply]
In the page Anpanator screenshotted above, I put the verify code in and click "Continue," and nothing happens.  Nothing at all.  It does seem like they insist on requiring a smartphone or mobile device which I do not, and will not, use.  For programming stuff, it's idiotic to even think about doing it on a smartphone.  If github doesn't provide an easy solution, I'll just dump it.  I'm only on one thing where I can make edits anyway, and it's not my own repo, and I hate the way github works, so it'll just be an excuse to get completely out, even though it will disappoint the person who owns the repo who wanted my input.

**Comment by bbkx226** [comment]


**Reply by Anpanator** [reply]
It really doesn't allow you to do that, at least not without first using SMS or Authenticator, see these screenshots.
![image](https://user-images.githubusercontent.com/11788825/224512659-fc7498d4-f457-4976-85e5-0d74e5d7a172.png)
![image](https://user-images.githubusercontent.com/11788825/224512679-34e0d289-69b5-4091-8999-0e9387a976ce.png)

**Reply by metux** [reply]
Note that quite recently they added showing the raw code instead of just the QR code image.
(maybe my own naggling has some share on it)
This makes it *a lot* easier.

**Comment by airtower-luna** [comment]
> I have a bunch of security keys here that I could use (and I always enroll at least 2 for backup), but not a mobile device.

You don't need a mobile device. The "authenticator app" can be anything that supports [TOTP according to RFC 6238](https://datatracker.ietf.org/doc/html/rfc6238). There are various desktop applications that do, [for example KeePassXC](https://keepassxc.org/docs/KeePassXC_UserGuide.html#_adding_totp_to_an_entry) (that's what Mozilla recommended for people who can't or don't want to use a mobile device when they started to require 2FA for add-on devs), and libraries for various languages in case you want to make your own.

When using a password manager that can store both your password *and* do TOTP, mind that having both in the same system reduces the security gain, though under most circumstances it's still better than *only* a password. Consider the security/convenience balance for yourself.

**Reply by Anpanator** [reply]
That might have to be the workaround (as you said, being on the same system isn't what I'd consider proper 2FA either). It seems like a silly restriction to put into place, especially when WebAuthn is actually supported.

Do you know if it's possible to completely remove TOTP as an 2FA method after the security keys are enrolled?

**Reply by airtower-luna** [reply]
I don't know, but I'd be surprised if they'd require it only initially. :sweat_smile:

**Reply by mezorian** [reply]
Thanks so much. You saved my live.. I don't have a smart phone and searched for a nice way around ugly SMS.

**Reply by SkybuckFlying** [reply]
TOTP is completely retarded, reason:

Algorithm uses unix time.

Unix time will break in 2036.

**Comment by DNin01** [comment, marked as answer]
GitHub did say in a blog post that [they were working on making it possible to use security keys as the primary second factor](https://github.blog/changelog/2022-11-21-updates-to-the-two-factor-authentication-setup-flow/), so if I'm right, this option will be available soon.

> In the future, you'll also find security keys there as an option for initial setup on supported devices and browsers.

Passkey support might also be interesting to you.

> Lastly, we’re already testing [passkeys](https://fidoalliance.org/passkeys/) internally, which we believe will combine ease of use with strong, phishing-resistant authentication. Keep an eye on this space for when this functionality is ready for you.

**Reply by Anpanator** [reply]
I wasn't aware of that blog post. Who knows when "in the future" is, but this answers my question. Thank you.

**Reply by MightyBOBcnc** [reply]
It has been 267 days since that blog post and there is zero indication that security keys will be allowed as a primary/sole 2FA method.

**Reply by Anpanator** [reply]
Yep, and it's doubly silly because I received the Email that 2FA will be enforced soon and in there it says I can choose one of 4 methods, including security keys, which is clearly not the case.

**Reply by Manuel-K** [reply]
I did get an email (that looked very much like a phishing attempt – so well done) and a banner on the top of the page, so I expect that I'll be locked out soon.

**Comment by mcdubhghlas** [comment]
> I understand GitHub wants to require 2FA soon and in principle I think this is a good thing.

It's not a good thing and 2FA isn't even particularly good for security, nor privacy.

Just force us to password reset every 45 days and if we login from a different IP address, make sure confirm in the email. Better alternative: Let us handle our own security.

So now I either have to buy a cell phone and pay for a monthly fee, or I have to install yet-more-software that I don't want on my computer. Fantastic. And for a website I don't have to use much. On the plus side, this would be a great point of concern to bring up getting away from github with my university. So, thanks for that if it works and I wish you well.

**Reply by Anpanator** [reply]
> It's not a good thing and 2FA isn't even particularly good for security, nor privacy.
> 
> Just force us to password reset every 45 days and if we login from a different IP address, make sure confirm in the email.

2FA is actually excellent for security. It's just not a silver bullet, like nothing in terms of security is. Security keys in particular not only make account take over nigh impossible even IF your account credentials somehow became known, they also protect against phishing, as the domain is part of the authentication protocol. And yes, even smart people, even security specialists, sometimes fall for phishing, because we're all human and phishing often exploits weaknesses we humans have.

It's been well documented what behavior will happen when you force people to reset their passwords regularly: They make minute changes to the password to get it over with (like adding a 1 in front every time, making the process nearly pointless) or worse yet, use much easier to guess passwords.

Using the IP is also not a great idea as that may change quite frequently (on mobile networks, when using VPNs or tor, etc.).

People are ultimately lazy, so I think to finally get security up a little bit, it's a great idea to require 2FA, so long as there are enough options. Not sure why you brought privacy up, that is not something that was ever meant to be addressed by 2FA to my knowledge.

**Reply by marcinn** [reply]
> 2FA is actually excellent for security. 

Forcing 2FA is not an excellent solution, same as forcing users to change their passwords. We have secure keys and strong passwords. That's pretty enough.

>  it's a great idea to require 2FA

No. You think it's a good idea, but in reality it's not.

**Reply by Anpanator** [reply]
> No. You think it's a good idea, but in reality it's not.

Would you care to elaborate on why it is not a good idea? It sounds to me you're simply unhappy about losing a minute amount of convenience, but I am happy to be corrected.

**Reply by marcinn** [reply]
> Would you care to elaborate on why it is not a good idea? 

You already know the answer. 

> It's been well documented what behavior will happen when you force people to reset their passwords regularly: They make minute  changes to the password to get it over with (like adding a 1 in front every time, making the process nearly pointless) or worse yet, use much easier to guess passwords.

**Forcing** someone to do something they don't want or don't need, is never a good idea. 


EDIT:

> It's just not a silver bullet, like nothing in terms of security is

Agreed. It's a tradeoff. For bank transfers I would go with 2FA (I prefer SMS over app, and printed codes over SMS). But I'm really sick of stupid browser fingerprint checking, which **forces me** to manage devices list on each update, and gives nothing in terms of security (there is 2FA already enabled).

I have no money on my GitHub account. Same as I have no money on my Apple account. But Apple ID requires me to accept the device and enter codes on every log-in. That's just insane.

**Reply by Anpanator** [reply]
That's a non-argument in terms of upping the rate at which accounts are taken over. I'm not arguing if it's **nice** to do this, but it will definitely help with securing accounts, even if it's a forced change (unlike forced password resets, which on average, will effectively lower security due to human behavior).
2FA can never be worse than no 2FA, however.

I can understand your misgivings about the change, it's just not what I was interested in. In terms of security, 2FA is definitely a good idea as far as I can tell.

**Reply by marcinn** [reply]
> That's a non-argument in terms of upping the rate at which accounts are taken over.

It is just a wrong assumption.  For 13yrs+ I had no takeover of my account.  It is protected with one strong, long, unchanged for years password. The ssh key is also old, but (AFAIR) I was updated to RSA just once. And literally no one will lose their life if my account gets hacked. 

**BUT** in case of doing business with GitHub as a tool, in case of risk of violating NDAs, I **would like** to enable 2FA. 

> 2FA can never be worse than no 2FA, however.

It's good to have possibility to use it when needed, not forced.

Hope you understand the difference now.

**Reply by Anpanator** [reply]
I've understood you the first time, you just haven't brought forward any arguments against 2FA that relate to it's effectiveness in terms of security. Again, I completely understand why you'd be unhappy with the change - it just has no bearing on whether or not 2FA improves security or not. The fact that you in particular have not had an account taken over doesn't change this.

**Reply by marcinn** [reply]
I will NOT share my phone number with any company other than CSP or bank. Only Apple forced me to do that - the first and last time. Sharing my phone is a security issue, and there can be potential data leak at GitHub. I'm not interested in possible phising attacks via SMS or phone calls just because someone in Microsoft will **FORCE** me to compromise **my** security.

**Reply by Anpanator** [reply]
According to the people answering here you do not have to use your phone for 2FA. And the accepted answer explains that GitHub will allow WebAuthn as the initial, primary 2FA too. Did you not check the answers?

**Reply by marcinn** [reply]
Yes, I did. I will not use the app, and I'm not interested in spending money on buying additional hardware just because you or someone else thinks that it is necessary. It is not necessary, as I have already proved.

**Reply by Anpanator** [reply]
You don't have to use a phone app or use additional hardware either. There's free open source software that allows you to do it (and if you have a password manager, you may even have it already, as mentioned above). If using even that is too much to ask of you, then I suppose GitHub won't be for you in the foreseeable future. That's actually what I thought might happen to me, so it's not like I can't relate.

**Reply by marcinn** [reply]
> There's free open source software that allows you to do it (and if you have a password manager, you may even have it already, as mentioned above)

Sorry, the only one password manager I have is my brain. It's blazing fast comparing to any 2FA method. (excl. ssh key auth)

> If using even that is too much to ask of you,  then I suppose GitHub won't be for you in the foreseeable future. 

Forced, not asked.  And yes, It won't be for me. I knew that Microsoft will destroy good product. As always.

**Reply by noembryo** [reply]
So, there is no way to access GitHub anymore from various devices without give Micro$oft your phone number.
Even using the extensions does not work since you may be on a device without them.
Great move! Thank you for the new problems that I never thought I could have...

**Reply by supervisitor** [reply]
You have now spoken the true reason: they want your mobile number! Like all those who think they have to offer security, even if you don't want it yourself.

It's ok to offer such features, so that those who think they need them can activate those. But making them mandatory has marketing reasons in every company: the email spam filters are too good, now they want to annoy you on Whatsapp, Telegram and Signal.

If there will be no possibility with brain, eMail, password to log in... then I am gone here (I know, I'm not important for them!). I am already annoyed when I have to search for the verification code in the email after entering the password. This is not the first portal and will probably not be the last.

**Reply by noloader** [reply]
GitLab. It will send access codes to an email address.

**Reply by metux** [reply]
Exactly. I'm currently looking for an alternative git hosting service that doesnt torture its users in such ridiculous ways, so i can move away all my repos and delete my account, forever.

**Reply by SenadUsername** [reply]
so are there any alternatives? Gitlab requres a phone number while registering, which means at some point there you also risk to loose access to your gitlab if you don't have access to your phone. I havent used bitbucket in a while, I dont know at what stage of data collection they are currently.

**Comment by yangsfang** [comment]
Yep. Seems really backwards. They would rather have you use SMS as a fallback rather than letting you use security keys through WebAuthn. Considering TOTP can be and was phished during highly public hacks of a certain security provider last year, I would think the priority is to let us use WebAuthn first, then enforce 2FA, and also get rid of the SMS option.

**Comment by eliot-akira** [comment]
I also had an issue with GitHub's 2FA seeming to require a mobile device.

On Linux there is an open-source application called Authenticator, which I found today and used successfully to generate TOTPs (time-based one-time passwords) for two-factor authentication. Note that in some aspects it may not be considered as secure as using another device, but at least it does not compromise your privacy.

[Authenticator: A Simple Open-Source App to Replace Authy on Linux](https://itsfoss.com/authenticator/)

https://flathub.org/apps/com.belmoussaoui.Authenticator

**Reply by realDragon11** [reply]
You're a lifesaver! I knew when this was announced I'd have to add a random 3rd party service vulnerability to my account, but I was expecting the official ones they linked to actually work on my machine.
Now I'm using some random internet person's code that I got linked in a complaint post to secure my account instead of a password I keep safe and never type in manually!

Isn't forcing security (or rather, extra hoops that add nothing because my phone is the most easily stolen thing I have and I regularly lose access to devices due to hardware issues) on users great?

EDIT was not expecting github to mandate I download a file with the recovery codes in plaintext with a very obvious name
wonder how many PCs have the 2fa recovery just laying in plaintext in their downloads folder now? probably about half of github users I'm betting

**Reply by GarthWilson** [reply]
eliot-akira, I just installed Authenticator to try.  It won't launch.

**Reply by zuzia-dev** [reply]
The latest version works well, installation by Flatpak.

**Comment by KK68HK** [comment]
My computer died dis morning.I don't know vats happen it's just died.And I have not manny to by a new one.I have been very sick for a long time so I don't have so much money.it vas very strange ting the GitHub vosent the same on my computer it was same before I don't know when I kan fix nuy computer I hop son.Kai

**Comment by supervisitor** [comment]
I am surprised, no one is questioning the introduction of 2FA in general?
We are in a discussion about the introduction of a mechanism for everyone, which actually only a small percentage, if any, requires.
Wouldn't it make more sense to discuss the rejection of the introduction of 2FA instead of accepting the introduction and burning hours to implement it in a way that is acceptable to everyone? Remember, only a few really need it, implementation for everyone else is a waste of time. 😎

**Reply by noembryo** [reply]
Totally agree, but the problem is that Microsoft is Micro$oft and apart of moving away there is nothing I see as an option.. 😔

**Reply by supervisitor** [reply]
My important projects are all in own/selfhosted repos/SVN systems. That's why I don't care to dismount my account here (unfollow, unstar, delete all,...). 😉
Everyone should always know a alternative, because on the internet everything can anytime be turned off. 😇 To name some of the alternatives (have a closer look and give them a try): Bitbucket, Codeberg, Frog Git... or host yourself.

**Reply by noloader** [reply]
GitLab. It will send access codes to an email address.

**Reply by metux** [reply]
Exactly. We dont need that cruft at all. Supply chain attacks are trivial to defeat by pure git mechs: signed tags.
Git was *designed* to be secure against untrusted or compromised hosts

**Reply by SenadUsername** [reply]
Gitlab request your phone number (or credit card info) during registration. I guess it is slightly better since you need a functioning phone only during the initial registration.

**Reply by KuroodoD** [reply]
I wasn't asked to input a phone number nor a credit card number when I made my GitLab account around 2 months ago, that's strange

**Reply by SenadUsername** [reply]
Maybe you have logged in with your google, github, bitbucket or salesforce account?  If you choose to create a new "standalone" account, then on the second step you are asked for phone or credit card info. I went to the process a couple minutes ago.

**Comment by Kitteh6660** [comment]
Wouldn't it be possible to, in the future, allow email 2FA as an option? Where if I try to log in, I receive the email containing single-use code. No need for phones, apps or security keys.

**Reply by supervisitor** [reply]
That's actually the security standard today.
I open a ticket (personal/0/#2312660) to get direct answer from GitHub support (all should do this! they have to get feedback how many don't need this 2FA bs) about a login process without mobile phone. They/GitHub/MS have concerns about mail accounts and wrote: "We cannot securely determine the security status of, for example, a user's email account, and so having 2FA required...". 😉

**Reply by noloader** [reply]
GitLab. It will send access codes to an email address.

**Reply by SenadUsername** [reply]
> "Wouldn't it be possible to, in the future, allow email 2FA as an option? Where if I try to log in, I receive the email containing single-use code. No need for phones, apps or security keys."

2FA could have easily been optional, but it is not. You could also have a perfectly secure account without 2FA but this is also not an provided option.

If you need a simple git provider, it would be best to migrate somewhere else until things get sorted. As a user that uses github for simple projects and testing I dont want the hassle of setting 2FA up for it and having to have a charged phone constantly on me, and having to set up log in alternatives for 2FA like one-time codes which i have to keep somewhere "safe" on my device since they are not easy to memorize. _Btw dont keep your 2FA app and one-time recovery codes on one device. It is best to carry at least two phones with you. One has to be the 2FA phone and the other has to be the phone that contains the one-time recovery codes._

**Comment by noloader** [comment]
Great job Microsoft! You DoS'd my account with your piss-poor planning.

I don't have a personal cell phone. I hate those god damn things. And I hate driving around kids using those damn things.

I have access to a landline, but you don't offer an option to use the house phone. Or I never got a call when I entered my house number. And you don't offer an option to provide access codes via email.

GitLab does a much better job at this crap. GitLab workflows are orders of magnitude better than what Microsoft has done.

In closing, great job assholes.

**Reply by metux** [reply]
We should move away. Just havent found a free alternative hoster yet

**Comment by kuroodo** [comment]
> "We cannot securely determine the security status of, for example, a user's email account, and so having 2FA required...". 

What a load of garbage. Neither can they determine the security status of, for example, a user's TOTP application and backup keys.

I also tried contacting support and was given the same easily refutable responses. But they do not care and are unwilling to compromise or accommodate.

I migrated my repos to GitLab and cancelled my copilot subscription. It was easy. So should you

**Comment by metux** [comment]
Does anyone know a tool for decoding the unreadable qr code into something human readable ?

**Reply by diomed** [reply]
any QR scanner can do that.

**Comment by severach** [comment]
2fa without a phone is easy. I found Authenticators for Firefox and Chrome. How do I get both browsers authorized? Each time I authorize a browser with the recovery key, the other browser dies.

Edit: Got it. For this extension must export and import to next browser so that all browsers are using the same key. Be sure to delete any extra copies of the same accounts so you don't get shown bogus response keys.

**Reply by noembryo** [reply]
2fa without a phone makes it really difficult to sign in using a customer's/client's machine.
Too much complication without a real reason..

**Reply by metux** [reply]
After intensive research, found a solution:

* zbarimg (from zbar-tools package) for decoding the QR code image into string (URL in totp scheme)
* oathtool for generating the tokens
* few lines of python for putting it all together (call zbarimg to decode, extract url params and call up oathtool)

notes:
* you'll explicitly have to enable TOTP mode in oathtool (--toptp)
* keys are in base32 (oathtool flag -b)

But:

github uses a *WEAK* key !

**Reply by metux** [reply]
Here's the python script that puts all together  - from decoding QR code to generating the current token:

https://gist.github.com/metux/ce43da58d9759019a0fbf4d25ccb1b00

**Reply by GarthWilson** [reply]
I installed Authenticator to give it a try.  It won't launch.  Doesn't run.  More and more, it's looking like when the deadline arrives in five days (I think it's 1/25/24), I'll be off of github.

**Comment by bathos** [comment]
I would be happy to use 2FA (MFA) via WebAuthn, which all major browsers already support and is better than “manual” 2FA, but this option seems to be missing. I do not own any mobile devices and the assumption that either everyone does or that no one else deserves to be able to continue using GitHub seems wild to me.

**Reply by Cheatoid** [reply]
Micro$oft ~~Microsoft~~ logic. However, [Authenticator.cc](https://chrome.google.com/webstore/detail/authenticator/bhghoamapcdpbohphigoooaddinpkbai) browser/Chrome extension seems to be working.

**Comment by arnvidhr** [comment]
I need e-mail based 2FA, nothing more. I'm against to use app for sure, against to save anything. **You may send it via email or SMS, nothing more or less. Nothing else!** We may stop to use github, if email or SMS 2FA options, without saving anything won't be available.

1. I won't install any apps on mobile devices, which possible to loose and to loose in this WAY access to the account.
2. **I won't save any codes**, to any storage, which can be lost the same as the access to the account, in such way. NO WAY!
3. **Give me an E-MAIL 2FA or remove this your mandatory!**
